### PR TITLE
Add a better toString to _ClientSocketException

### PR DESCRIPTION
**Pull request by brianquinlan**
Also explain how to catch `SocketException`.

Fixes https://github.com/dart-lang/http/issues/930

- Thanks for your contribution! Please replace this bullet-point with a description of what this PR is changing or adding and why; list any relevant issues, and review the contribution guidelines below

---

- [X] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

#### Contribution guidelines:

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

(note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback)

### CODE REVIEW

File: pkgs/http/lib/src/client.dart [hunk 1]
```diff
@@ -25,6 +25,10 @@ import 'streamed_response.dart';
 /// [http.head], [http.get], [http.post], [http.put], [http.patch], or
 /// [http.delete] instead.
 ///
+/// All methods will emit a [ClientException] if there is a transport-level
+/// failure when communication with the server, e.g., if the server could not
```
Comment (natebosch) [inline pkgs/http/lib/src/client.dart:29]: [nit] Avoid using "e.g."

https://dart.dev/effective-dart/documentation#avoid-abbreviations-and-acronyms-unless-they-are-obvious

File: pkgs/http/lib/src/io_client.dart [hunk 2]
```diff
@@ -40,9 +41,33 @@ class _ClientSocketException extends ClientException
...
+  @override
+  String toString() => 'ClientException with $cause, uri=$uri';
 }
 
-/// A `dart:io`-based HTTP client.
+/// A `dart:io`-based HTTP [Client].
+///
+/// If there is a socket-level failure when communicating with the server
+/// (e.g. if the server could not be reached), [IOClient] will emit a
+/// [ClientException] that is also implements [SocketException]. This allows
```
Comment (natebosch) [inline pkgs/http/lib/src/io_client.dart:53]:
```suggestion
/// [ClientException] that also implements [SocketException]. This allows
```

**Review by natebosch — APPROVED** [review]
Need to bump to a `-wip` version in the pubspec. May also be worth a changelog entry.

File: pkgs/http/lib/src/client.dart (hunk 1 above)
Comment (brianquinlan) [reply pkgs/http/lib/src/client.dart:29]: Done

File: pkgs/http/lib/src/io_client.dart (hunk 2 above)
Comment (brianquinlan) [reply pkgs/http/lib/src/io_client.dart:53]: Done

**Review by natebosch — APPROVED** [review]

**Comment by brianquinlan** [comment]
> Need to bump to a `-wip` version in the pubspec. May also be worth a changelog entry.

Done.

**Pull request merged**